### PR TITLE
Feature/add metadata protos

### PR DESCRIPTION
**Pull request by iramiller**
## Description

This change set for the metadata module includes the basic proto definitions and a minimal set of associated go types to achieve a working build.

closes: #26 

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [x] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/provenance-io/provenance/blob/master/CONTRIBUTING.md#pr-targeting))
- [x] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Wrote unit and integration [tests](https://github.com/provenance-io/provenance/blob/master/CONTRIBUTING.md#testing)
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`)
- [ ] Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [x] Re-reviewed `Files changed` in the Github PR explorer
- [x] Review `Codecov Report` in the comment section below once CI passes

### CODE REVIEW

File: proto/provenance/metadata/v0/contract.proto [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+import "gogoproto/gogo.proto";
+import "provenance/metadata/v0/spec.proto";
+import "provenance/metadata/v0/types.proto";
+
+import "google/protobuf/timestamp.proto";
+
+message Recital {
+  types.PartyType                      signer_role = 1;
+  types.SigningAndEncryptionPublicKeys signer      = 2;
+  bytes                                address     = 3;
```
Comment (dpederson-figure) [inline proto/provenance/metadata/v0/contract.proto:18]: what's best practice in Cosmos now, bytes or strings for addresses?  No need to change, just curious.

File: proto/provenance/metadata/v0/scope.proto [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+message Scope {
+  // Unique ID for building up ProvenanceReferences
+  types.UUID uuid = 1;
+  // Required parties on this scope.  These are union with any contract specific recital requirements
+  repeated contract.Recital parties = 2;
+  // Each contract memorialized in this scope is stored in this list of contract_group.  The list of groups is keyed
+  // as a hashmap based on the contract_spec_ref as a unique key.
+  repeated scope.RecordGroup record_group = 3;
+  // Last event that resulted in scope changes.
+  scope.Event last_event = 4;
```
Comment (dpederson-figure) [inline proto/provenance/metadata/v0/scope.proto:41]: will need value_owner here at some point - with whatever type is standard for representing addresses in proto.

File: proto/provenance/metadata/v0/contract.proto (hunk 1 above)
Comment (iramiller) [reply proto/provenance/metadata/v0/contract.proto:18]: its supposed to be strings everywhere ... but here in `v0` proto I am trying to follow the existing Figure spec

File: proto/provenance/metadata/v0/scope.proto (hunk 2 above)
Comment (iramiller) [reply proto/provenance/metadata/v0/scope.proto:41]: should have that in the `v1` protos ...

File: proto/provenance/metadata/v1/scope.proto [hunk 3]
```diff
@@ -0,0 +1,234 @@
...
+
+
+*/
+
+// Scope defines a root reference for a collection of records owned by one or more parties.
+message Scope {
+  option (gogoproto.goproto_stringer) = false;
+
+  // Unique ID for this scope.  Implements sdk.Address interface for use where addresses are required in Cosmos
+  bytes scope_id = 1 [
```
Comment (dpederson-figure) [inline proto/provenance/metadata/v1/scope.proto:74]: cool, but won't this be a huge change for p8e?  Are they aware that scope ID will be changing from UUID to address?

**Review by dpederson-figure — APPROVED** [review]
Looks like a good start to me.

File: proto/provenance/metadata/v1/scope.proto (hunk 3 above)
Comment (iramiller) [reply proto/provenance/metadata/v1/scope.proto:74]: The response types for messages in queries will need to return the UUID.  Same for appropriate events.  The address class has some extensive methods/tests/mappings to make that easy to accomplish.

**Review by mtps — APPROVED** [review]

**Pull request merged**